phoenix-core/src/main/java/org/apache/phoenix/coprocessor/GlobalIndexRegionScanner.java

```diff
@@ -0,0 +1,161 @@
+ * limitations under the License.
+ */
+package org.apache.phoenix.coprocessor;
+
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.Mutation;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.hadoop.hbase.regionserver.Region;
+import org.apache.hadoop.hbase.regionserver.RegionScanner;
+import org.apache.hadoop.hbase.regionserver.ScanInfoUtil;
+import org.apache.hadoop.hbase.util.Pair;
+import org.apache.phoenix.hbase.index.ValueGetter;
+import org.apache.phoenix.hbase.index.covered.update.ColumnReference;
+import org.apache.phoenix.hbase.index.parallel.TaskBatch;
+import org.apache.phoenix.hbase.index.parallel.TaskRunner;
+import org.apache.phoenix.hbase.index.table.HTableFactory;
+import org.apache.phoenix.index.IndexMaintainer;
+import org.apache.phoenix.mapreduce.index.IndexVerificationResultRepository;
+
+import java.util.List;
+import java.util.Map;
+
+public abstract class GlobalIndexRegionScanner extends BaseRegionScanner {
+
+    public static final String NUM_CONCURRENT_INDEX_VERIFY_THREADS_CONF_KEY = "index.verify.threads.max";
+    public static final int DEFAULT_CONCURRENT_INDEX_VERIFY_THREADS = 17;
+    public static final String INDEX_VERIFY_ROW_COUNTS_PER_TASK_CONF_KEY = "index.verify.threads.max";
+    public static final int DEFAULT_INDEX_VERIFY_ROW_COUNTS_PER_TASK = 2048;
+    public static final String NO_EXPECTED_MUTATION = "No expected mutation";
+    public static final String ACTUAL_MUTATION_IS_NULL_OR_EMPTY = "actualMutationList is null or empty";
+    public static final String ERROR_MESSAGE_MISSING_INDEX_ROW_BEYOND_MAX_LOOKBACK = "Missing index row beyond maxLookBack";
+    public static final String ERROR_MESSAGE_MISSING_INDEX_ROW = "Missing index row";
+
+    protected long pageSizeInRows = Long.MAX_VALUE;
+    protected int rowCountPerTask;
+    protected boolean hasMore;
+    protected int maxBatchSize;
+    protected UngroupedAggregateRegionObserver.MutationList mutations;
+    protected byte[] indexMetaData;
+    protected Scan scan;
+    protected RegionScanner innerScanner;
+    protected Region region;
+    protected IndexMaintainer indexMaintainer;
+    protected Table indexHTable = null;
+    protected Map<byte[], Put> indexKeyToDataPutMap;
+    protected TaskRunner pool;
+    protected TaskBatch<Boolean> tasks;
+    protected String exceptionMessage;
+    protected HTableFactory hTableFactory;
+    protected int indexTableTTL;
+    protected long maxLookBackInMills;
+    protected IndexToolVerificationResult verificationResult;
+    protected IndexVerificationResultRepository verificationResultRepository;
+
+    public GlobalIndexRegionScanner(RegionScanner delegate) {
+        super(delegate);
```

Comment:
This constructor can initialize the attributes that are defined in this class. This will allow you to remove the common code from the subclass constructors.

Reply:
Oh, yeah that's right!